crates/pipeline-types/src/program_schema.rs

```diff
@@ -0,0 +1,96 @@
+use serde::{Deserialize, Serialize};
+use utoipa::ToSchema;
+
+#[cfg(feature = "testing")]
+use proptest::{collection::vec, prelude::any};
+
+/// Returns canonical form of a SQL identifier:
+/// - If id is _not_ quoted, then it is interpreted as a case-insensitive
+///   identifier and is converted to the lowercase representation
+/// - If id _is_ quoted, then it is a case-sensitive identifier and is returned
+///   as is, without quotes.
+pub fn canonical_identifier(id: &str) -> String {
+    if id.starts_with('"') && id.ends_with('"') && id.len() >= 2 {
+        id[1..id.len() - 1].to_string()
+    } else {
+        id.to_lowercase()
+    }
+}
+
+/// A struct containing the tables (inputs) and views for a program.
+///
+/// Parse from the JSON data-type of the DDL generated by the SQL compiler.
+#[derive(Serialize, Deserialize, ToSchema, Debug, Eq, PartialEq, Clone)]
+#[cfg_attr(feature = "testing", derive(proptest_derive::Arbitrary))]
+pub struct ProgramSchema {
+    #[cfg_attr(
+        feature = "testing",
+        proptest(strategy = "vec(any::<Relation>(), 0..2)")
+    )]
+    pub inputs: Vec<Relation>,
+    #[cfg_attr(
+        feature = "testing",
+        proptest(strategy = "vec(any::<Relation>(), 0..2)")
+    )]
+    pub outputs: Vec<Relation>,
+}
+
+/// A SQL table or view. It has a name and a list of fields.
+///
+/// Matches the Calcite JSON format.
```

Comment:
Calcite or SQL compiler?

Reply:
I didn't write this comment, but I'm pretty sure it's the SQL compiler.